contracts/IFTieredSale.sol

```diff
@@ -208,101 +207,132 @@ contract IFTieredSale is ReentrancyGuard, AccessControl, IFFundable {
+                _updatePromoCodeRewards(_promoCode, price * _amount, _tierId);
+            } else {
+                _updateWalletPromoCodeRewards(_walletPromoCode, price * _amount);
+            }
+        }
+
+        executePurchase(_tierId, _amount, price, finalPromoCode);
     }
 
-    function whitelistedPurchaseInTierWithWalletCode(
+    function signedPurchaseInTierWithCode(
         string memory _tierId,
         uint256 _amount,
-        bytes32[] calldata _merkleProof,
-        address _walletPromoCode,
-        uint256 _allocation
+        uint256 _allocation,
+        bytes calldata signature,
+        string memory _promoCode,
+        address _walletPromoCode
     ) public {
-        // Ensure promo codes are allowed for the tier and the promo code is valid
-        require(tiers[_tierId].allowWalletPromoCode, "Promo code is not allowed for this tier");
-        require(msg.sender != _walletPromoCode, "Cannot purchase with own wallet address promo code");
-        // the promo code wallet address has to purchase at least one node
-        require(validateWalletPromoCode(_walletPromoCode), "Promo code address has not purchased a node");
-        string memory promoCode = addressToString(_walletPromoCode);
-        // no need to validate address promo code at purchase
-        bytes32 tierWhitelistRootHash = tiers[_tierId].whitelistRootHash;
-        if (tierWhitelistRootHash != bytes32(0)) {
-            require(checkTierWhitelist(_tierId, msg.sender, _merkleProof, _allocation), "Invalid proof");
-            require(purchasedAmountPerTier[_tierId][msg.sender] + _amount <= _allocation, "Purchase exceeds allocation");
-        }
+        require(tiers[_tierId].requireSignature, "Use whitelisted purchase");
+        bytes32 messageHash = keccak256(abi.encodePacked(msg.sender, _tierId, _allocation));
 
-        uint8 discount = calculateDiscount(promoCode);
-        uint256 discountedPrice = tiers[_tierId].price * (100 - discount) / 100;  // in gwei
-        executePurchase(_tierId, _amount, discountedPrice, promoCode);
+        bytes32 message = ECDSA.toEthSignedMessageHash(messageHash);
 
-        codePurchaseAmount[promoCode] += discountedPrice * _amount;
-        _updateWalletPromoCodeRewards(_walletPromoCode, discountedPrice * _amount);
-    }
+        address signer = ECDSA.recover(message, signature);
 
-    function calculateDiscount(string memory _promoCode) internal view returns (uint8) {
-        uint8 discount;
-        if (_isWalletPromoCode(_promoCode)) {
-            discount = addressPromoCodeDiscountPercentage; // Fixed discount for address-based promo codes
-        } else {
-            discount = promoCodes[_promoCode].discountPercentage; // Variable discount for other promo codes
+        // the message has to be signed by operator
+        require(hasRole(OPERATOR_ROLE, signer), "Invalid signature");
+
+        require(getTotalPurchasedAmount(msg.sender) + _amount <= _allocation, "Purchase exceeds allocation");
```

Comment:
Allocation changed to maxPaymentSent
GetTotalPurchasedAmoutn changed to GetTotalPaymentSent
_amount changed to amount * price in this tier